File: 086_partition_list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode partition(ListNode head, int x) {
        // create two list then combine together
        ListNode dumm1 = new ListNode(0);
        ListNode dumm2 = new ListNode(0);
        ListNode l1 = dumm1;
        ListNode l2 = dumm2;
        ListNode curr = head;

        while (curr != null) {
            if (curr.val < x) {
                l1.next = curr;
                l1 = l1.next;
            } else {
                l2.next = curr;
                l2 = l2.next;
            }
            curr = curr.next;
        }
        // concate
        l1.next = dumm2.next;
        l2.next = null;
        return dumm1.next;
    }
}

File: python/086_partition_list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def partition(self, head: ListNode, x: int) -> ListNode:
        curr_small = dummy_small = ListNode(0)
        curr_large = dummy_large = ListNode(0)

        while head:
            if head.val < x:
                curr_small.next = head
                curr_small = curr_small.next
            else:
                curr_large.next = head
                curr_large = curr_large.next
            head = head.next

        curr_large.next = None
        curr_small.next = dummy_large.next
        return dummy_small.next
